treewide: replace `stdenv.is` with `stdenv.hostPlatform.is`

Artturin opened the pull request:
# Approve or thumbs up to show that you approve of the idea, or request changes on the first file to discuss why you don't think it's a good idea.

Join https://matrix.to/#/#stdenv:nixos.org to discuss

I'll reset to master and rerun the commands before merge.
Further replacements can be done in the PR where the warnings are added.
The commits can be added to `.git-blame-ignore-revs` in a separate PR

treewide: replace `stdenv.is` with `stdenv.hostPlatform.is`

In preparation for the deprecation of `stdenv.isX`.

These shorthands are not conducive to cross-compilation because they
hide the platforms.

Darwin might get cross-compilation for which the continued usage of `stdenv.isDarwin` will get in the way

One example of why this is bad and especially affects compiler packages
https://github.com/NixOS/nixpkgs/pull/343059

There are too many files to go through manually but a treewide should
get users thinking when they see a `hostPlatform.isX` in a place where it
doesn't make sense.

```
fd --type f "\.nix" | xargs sd --fixed-strings "stdenv.is" "stdenv.hostPlatform.is"
fd --type f "\.nix" | xargs sd --fixed-strings "stdenv'.is" "stdenv'.hostPlatform.is"
fd --type f "\.nix" | xargs sd --fixed-strings "clangStdenv.is" "clangStdenv.hostPlatform.is"
fd --type f "\.nix" | xargs sd --fixed-strings "gccStdenv.is" "gccStdenv.hostPlatform.is"
fd --type f "\.nix" | xargs sd --fixed-strings "stdenvNoCC.is" "stdenvNoCC.hostPlatform.is"
fd --type f "\.nix" | xargs sd --fixed-strings "inherit (stdenv) is" "inherit (stdenv.hostPlatform) is"
fd --type f "\.nix" | xargs sd  "buildStdenv.is" "buildStdenv.hostPlatform.is"
fd --type f "\.nix" | xargs sd  "effectiveStdenv.is" "effectiveStdenv.hostPlatform.is"
fd --type f "\.nix" | xargs sd  "originalStdenv.is" "originalStdenv.hostPlatform.is"
```

The rebuilds are due to `src ./.` or similar

![image](https://github.com/user-attachments/assets/4f9e8397-f455-403d-806b-e8ec4893f6d7)

https://gist.github.com/Artturin/4cc1e73fc612cb7a79459a9bb6bc27a3

emilazy commented:
I keep slipping up and using this, so +1 for deprecation.

Artturin commented:
Will fix editorconfig files in a separate PR

```
$ fd --type f "\.nix" | nix-shell -p editorconfig-checker --run 'xargs -r editorconfig-checker -disable-indent-size'
pkgs/data/icons/apple-cursor/default.nix:
	34: Trailing whitespace

1 errors found
pkgs/applications/science/math/sage/dist-tests.nix:
	10: Wrong indent style found (tabs instead of spaces)
	11: Wrong indent style found (tabs instead of spaces)
	12: Wrong indent style found (tabs instead of spaces)
	13: Wrong indent style found (tabs instead of spaces)
pkgs/build-support/rust/build-rust-crate/log.nix:
	52: Wrong indent style found (tabs instead of spaces)
	55: Wrong indent style found (tabs instead of spaces)
	56: Wrong indent style found (tabs instead of spaces)

7 errors found
pkgs/build-support/dotnet/dotnetenv/build-solution.nix:
	36: Wrong indent style found (tabs instead of spaces)
	37: Wrong indent style found (tabs instead of spaces)
	38: Wrong indent style found (tabs instead of spaces)
	39: Wrong indent style found (tabs instead of spaces)
	40: Wrong indent style found (tabs instead of spaces)
	41: Wrong indent style found (tabs instead of spaces)
	43: Wrong indent style found (tabs instead of spaces)
	44: Wrong indent style found (tabs instead of spaces)
	45: Wrong indent style found (tabs instead of spaces)
	50: Wrong indent style found (tabs instead of spaces)
	51: Wrong indent style found (tabs instead of spaces)
	53: Wrong indent style found (tabs instead of spaces)
	54: Wrong indent style found (tabs instead of spaces)
	55: Wrong indent style found (tabs instead of spaces)
	56: Wrong indent style found (tabs instead of spaces)
	57: Wrong indent style found (tabs instead of spaces)
	58: Wrong indent style found (tabs instead of spaces)
	60: Wrong indent style found (tabs instead of spaces)
	67: Wrong indent style found (tabs instead of spaces)
	68: Wrong indent style found (tabs instead of spaces)
pkgs/build-support/dotnet/dotnetenv/wrapper.nix:
	33: Wrong indent style found (tabs instead of spaces)
	34: Wrong indent style found (tabs instead of spaces)
	35: Wrong indent style found (tabs instead of spaces)
	36: Wrong indent style found (tabs instead of spaces)
	37: Wrong indent style found (tabs instead of spaces)
	38: Wrong indent style found (tabs instead of spaces)
	40: Wrong indent style found (tabs instead of spaces)
	41: Wrong indent style found (tabs instead of spaces)
	42: Wrong indent style found (tabs instead of spaces)
	51: Wrong indent style found (tabs instead of spaces)
	60: Wrong indent style found (tabs instead of spaces)
	61: Wrong indent style found (tabs instead of spaces)

32 errors found
```

circl-lastname commented:
I don't thing there's a need for another PR? Just amend this one with a fixing commit

Artturin commented:
> I don't thing there's a need for another PR? Just amend this one with a fixing commit

They're unrelated so I think a separate is better when the PR is already this big. https://github.com/NixOS/nixpkgs/pull/341423

wolfgangwalther reviewed (commented):
I confirmed the first commit matches the command mentioned.

There are two more patterns maybe worth replacing:
- `stdenv'.is`
- `stdenvNoCC.is`

Artturin commented:
> I confirmed the first commit matches the command mentioned.
> 
> There are two more patterns maybe worth replacing:
> 
>     * `stdenv'.is`
> 
>     * `stdenvNoCC.is`

Done and commands added to commit msg

Artturin commented:
Redid the commands with `--fixed-strings` and did `"inherit (stdenv) is" -> "inherit (stdenv.hostPlatform) is"`

Artturin commented:
The rebuilds are due to `src ./.` or similar

![image](https://github.com/user-attachments/assets/4f9e8397-f455-403d-806b-e8ec4893f6d7)

https://gist.github.com/Artturin/4cc1e73fc612cb7a79459a9bb6bc27a3

Artturin commented:
Undrafting to get reviews from probably all the `codeowners` in nixpkgs.

To keep the emails to a minimum either approve or thumbs up the PR if the idea is okay, or if it needs changes then let's discuss in https://matrix.to/#/#stdenv:nixos.org

K900 commented:
Idea is OK, list of merge conflicts is scary. Maybe do this stepwise?

roberth commented on lib/tests/test-with-nix.nix:
```diff
@@ -32,7 +32,7 @@ pkgs.runCommand "nixpkgs-lib-tests-nix-${nix.version}" {
     nix
     pkgs.gitMinimal
-  ] ++ lib.optional pkgs.stdenv.isLinux pkgs.inotify-tools;
+  ] ++ lib.optional pkgs.stdenv.hostPlatform.isLinux pkgs.inotify-tools;
```
```suggestion
  ] ++ lib.optional pkgs.stdenv.buildPlatform.isLinux pkgs.inotify-tools;
```

This probably applies to most conditionals for any `nativeBuildInputs`.

emilazy replied:
Let’s maybe fix these patterns in another PR, as it doesn’t match the current behaviour and would disrupt the primary metric (almost no rebuilds) we can judge this PR’s safety by? At least now it will be obviously wrong.

roberth replied:
Good point, but maybe leave the bad ones as is, so they're easy to update in the follow-up PR? Or I guess this PR's diff could be used for that purpose. It's a bit more hidden, but about as good.

SGTM

emilazy replied:
I think it would require a smarter script than a simple text substitution to diagnose and avoid these cases, so I’d personally prefer to avoid holding this up over that.

sternenseemann replied:
`hostPlatform` may just as easily be correct for `nativeBuildInputs`. I'd just mechanistically replace them. Mistakes are mistakes before *and* after, they are just more obvious after.

I'd guess in most cases it also sort of doesn't matter since the derivations don't cross compile to begin with or (as is the case here) don't really make sense to cross compile.

emilazy commented:
Let’s just rebase, wait for the ofborg rebuilds number, and merge. Best to get this in before 24.11 because of backports.

Artturin committed:
treewide: replace `stdenv.is` with `stdenv.hostPlatform.is`

In preparation for the deprecation of `stdenv.isX`.

These shorthands are not conducive to cross-compilation because they
hide the platforms.

Darwin might get cross-compilation for which the continued usage of `stdenv.isDarwin` will get in the way

One example of why this is bad and especially affects compiler packages
https://www.github.com/NixOS/nixpkgs/pull/343059

There are too many files to go through manually but a treewide should
get users thinking when they see a `hostPlatform.isX` in a place where it
doesn't make sense.

```
fd --type f "\.nix" | xargs sd --fixed-strings "stdenv.is" "stdenv.hostPlatform.is"
fd --type f "\.nix" | xargs sd --fixed-strings "stdenv'.is" "stdenv'.hostPlatform.is"
fd --type f "\.nix" | xargs sd --fixed-strings "clangStdenv.is" "clangStdenv.hostPlatform.is"
fd --type f "\.nix" | xargs sd --fixed-strings "gccStdenv.is" "gccStdenv.hostPlatform.is"
fd --type f "\.nix" | xargs sd --fixed-strings "stdenvNoCC.is" "stdenvNoCC.hostPlatform.is"
fd --type f "\.nix" | xargs sd --fixed-strings "inherit (stdenv) is" "inherit (stdenv.hostPlatform) is"
fd --type f "\.nix" | xargs sd --fixed-strings "buildStdenv.is" "buildStdenv.hostPlatform.is"
fd --type f "\.nix" | xargs sd --fixed-strings "effectiveStdenv.is" "effectiveStdenv.hostPlatform.is"
fd --type f "\.nix" | xargs sd --fixed-strings "originalStdenv.is" "originalStdenv.hostPlatform.is"
```

Artturin committed:
treewide: reformat files which need reformatting after

`treewide: replace stdenv.is with stdenv.hostPlatform.is`

emilazy commented:
Reproduced the commits locally, ofborg has reported low build numbers on previous rebases, seems impossible to race ofborg and merge conflicts. Merging now and will take responsibility for the revert if ofborg throws up a scary result.
